### PR TITLE
BCDA-4084 - Migrate from GORM to raw SQL

### PR DESCRIPTION
**Pull request by mtrang1263**
### Fixes [BCDA-4084](https://jira.cms.gov/browse/BCDA-4084)

The team has decided to move off of GORM in favor of writing our own SQL queries using [go-sqlbuilder](https://github.com/huandu/go-sqlbuilder). By running our own queries, we hope to be able to achieve the following:

1. Avoid issue with "unexpected" queries generated by GORM. We've encountered [regressions](https://github.com/CMSgov/bcda-app/pull/614) with the library generating queries that we were not expecting.
2. Have a better understanding on how we are querying our data by easily searching for the clauses used in our codebase. Since our usage of GORM was proliferated throughout the code base, we were not confident on our query patterns and could not easily determine how we can evolve our queries and schema.
 

### Proposed Changes
1. Replace existing usages of GORM with an equivalent SQL query.
2. Centralize our queries behind an interface to allow us to continue to decouple parts of our application from the data layer.

### Change Details
1. Update our repository.go interface and implementation to include all of the queries needed in the main bcda-app.
2. Introduce repository.go interface and implementation to include all of the queries needed in bcda-worker.
3. For main code paths, force all interactions to go through the repository interface. This allows us to control what database actions are occurring in the made code path.
4. Remove gorm.Model from the various models. Explicitly adding ID for all models and createdAt and updatedAt where needed.

### Security Implications
- [x] new software dependencies
1. Introduced SQL builder library: https://github.com/huandu/go-sqlbuilder
2. Remove GORM dependency

- [ ] security controls or supporting software altered
- [ ] new data stored or transmitted
- [ ] security checklist is completed for this change
- [ ] requires more information or team discussion to evaluate security implications
- [x] no PHI/PII is affected by this change

### Acceptance Validation
1. CI passes. **DONE**
2. Deploy to lower environment dev/test to confirm functionality. Will plan to be more aggressive with manual testing in these environments since we completely changed are we're accessing our data from postgres. **DONE**
        * Ran manual requests on large ACO and verified E2E that I am able to download the data.
3. Deploy to sandbox and let it soak for a period of time. This will vet out the code changes with real user traffic.

### Feedback Requested

1. Validate that GORM queries are replaced with the equivalent SQL queries correctly.

### Follow-up work
1. https://jira.cms.gov/browse/BCDA-4134 -- Remove cclf_beneficiary_xrefs table
2. https://jira.cms.gov/browse/BCDA-4135 -- Remove unused columns from various tables.

### CODE REVIEW

File: unit_test.sh [hunk 1]
```diff
@@ -13,7 +13,19 @@ mkdir -p test_results/latest
 DB_HOST_URL=${DB}?sslmode=disable
 TEST_DB_URL=${DB}/bcda_test?sslmode=disable
 echo "Running unit tests and placing results/coverage in test_results/${timestamp} on host..."
-DATABASE_URL=$TEST_DB_URL QUEUE_DATABASE_URL=$TEST_DB_URL gotestsum --junitfile test_results/${timestamp}/junit.xml -- -race ./... -coverprofile test_results/${timestamp}/testcoverage.out 2>&1 | tee test_results/${timestamp}/testresults.out
+
+# Avoid the db/migrations package since it only contains test code
+PACKAGES_TO_COVER=$(go list ./... | egrep -v 'test|mock|db/migrations' | tr "\n" "," | sed -e 's/,$//g')
+# Leverage the coverpkg flag to allow for coverage of packages even if there are tests outside of its package.
```
Comment (adamayd) [inline unit_test.sh:19]: Is this `coverprofile` instead of `coverpkg`?

File: bcda/models/postgres/postgrestest/postgrestest.go [hunk 2]
```diff
@@ -0,0 +1,364 @@
+// Package postgrestest provides CRUD utilities for the postgres database.
+// These utilities allow the caller to modify the database in ways that we
+// wouldn't want to permit in the main code path.
+// To protect against usage in non-test code, all methods should accept
+// a *testing.T struct.
```
Comment (adamayd) [inline bcda/models/postgres/postgrestest/postgrestest.go:5]: Love this, if I didn't already say that somewhere else.  It's been a long review :)

**Review by adamayd — DISMISSED** [review]
I believe all my comments are questions more than changes.  I think I even answered one but can't find it to remove it.  I wish there was a better naming convention for Go variables, including the interfaces.  Kudos on the amount of work completed in this PR.

File: bcda/api/requests.go [hunk 3]
```diff
@@ -46,14 +54,10 @@ func NewHandler(resources []string, basePath string) *Handler {
 
 	h := &Handler{}
 
-	db := database.GetGORMDbConnection()
-	dbc, err := db.DB()
```
[the remark this answers is not in the document]
Comment (adamayd) [reply bcda/api/requests.go:50]: Found it!  I hadn't made it to `connections.go` yet before asking this.

File: bcda/api/v1/api.go [hunk 4]
```diff
@@ -108,15 +110,23 @@ func BulkGroupRequest(w http.ResponseWriter, r *http.Request) {
...
-	var job models.Job
-	err := db.First(&job, jobID).Error
+	job, jobKeys, err := h.Svc.GetJobAndKeys(context.Background(), uint(jobID))
 	if err != nil {
-		log.Print(err)
+		log.Error(err)
 		oo := responseutils.CreateOpOutcome(responseutils.Error, responseutils.Exception, responseutils.DbErr, "")
+		// NOTE: This is a catch all and may not necessarily mean that the job was not found.
+		// So returning a StatusNotFound may be a misnomer
 		responseutils.WriteError(oo, w, http.StatusNotFound)
```
[the remark this answers is not in the document]
Comment (mtrang1263) [reply bcda/api/v1/api.go:130]: Let me add it to the brainstorming session. I'll document these cases and we'll discuss.

File: bcda/bcdacli/cli.go [hunk 5]
```diff
@@ -762,3 +782,16 @@ func cloneCCLFZip(src, dst string) error {
 
 	return nil
 }
+
+func genPublicKey(publicKey io.Reader) (string, error) {
```
[the remark this answers is not in the document]
Comment (mtrang1263) [reply bcda/bcdacli/cli.go:786]: I'll create a ticket in our TBG to do it!

File: unit_test.sh (hunk 1 above)
Comment (mtrang1263) [reply unit_test.sh:19]: `coverprofile` writes the coverage numbers out to a file.

`coverpkg` allows tests in other packages to still count as "tested" for the listed packages. For example, our postgres package (before) did not have that many unit tests to verify the query functionality. However, we have E2E tests that use the actualy postgres repository. I was thinking it might be a good idea to apply coverage numbers of *any* of our tests called the function, but didn't want to do it w/o more team discussion.

Left in the unit_test.sh just so we have any easy hook if we decide to go for it.

Details here: https://golang.org/pkg/cmd/go/internal/test/

```
-coverpkg pattern1,pattern2,pattern3
        Apply coverage analysis in each test to packages matching the patterns.
        The default is for each test to analyze only the package being tested.
        See 'go help packages' for a description of package patterns.
        Sets -cover.
```

**Review by mtrang1263 — COMMENTED** [review]
With the latest barrage of unit tests on the  SQL queries, we moved coverage up.

Before:
```
bcda/models/postgres (148ms) (coverage: 34.2% of statements)
∅  bcdaworker/repository/postgres [means no coverage]
```

After:
```
bcda/models/postgres (251ms) (coverage: 89.5% of statements)
bcdaworker/repository/postgres (163ms) (coverage: 89.7% of statements)
```

With these unit tests, we would've caught the bug in the UPDATE calls w/o  a WHERE clause.

It should be noted that even though we had low coverage numbers, our E2E tests that were using the postgres repository verified most of the calls.

The unit-tests are more targeted and can verify edge-cases as well as details that are not relevant to the business layer. Ex: update cclf file should only update the single cclf file, not all of them.

File: bcda/api/requests.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mtrang1263) [reply bcda/api/requests.go:50]: Ah sorry, missed this comment. Now that we have direct access to the `*sql.DB` the getter does not return an error.

**Review by SMLuthi — APPROVED** [review]
LGTM! This looks fantastic and is quite the lift. Only found one small spelling error (and its in a comment so whatever). Especially like the `repository.go` changes with the addition of the interfaces and the addition of the `postgrestest.go`. Great job on this.

**Review by SMLuthi — APPROVED** [review]

**Review by adamayd — APPROVED** [review]
LGTM
